### PR TITLE
Can run custom tasks in Gobblin.

### PR DESCRIPTION
**Pull request by ibuenros**
This PR allows running custom tasks in Gobblin. 

To use custom tasks, implement a `TaskFactory` and use `TaskUtils.setTaskFactoryClass(WorkUnitState, Class<? extends TaskFactory>` in the `Source`. Tasks need to implement `TaskIFace`.

### CODE REVIEW

**Comment by ibuenros** [comment]
@ydai1124 @htran1 can you take a look?

File: gobblin-runtime/src/main/java/gobblin/runtime/SafeDatasetCommit.java [hunk 1]
```diff
@@ -129,27 +150,62 @@ public Void call()
...
+  }
+
+  @Data
+  private static class TaskFactoryWrapper {
+    private final TaskFactory taskFactory;
 
-    // Set the dataset state to COMMITTED upon successful commit
-    datasetState.setState(JobState.RunningState.COMMITTED);
+    public boolean equals(Object other) {
+      if (other instanceof TaskFactoryWrapper) {
```
Comment (ydai1124) [inline gobblin-runtime/src/main/java/gobblin/runtime/SafeDatasetCommit.java]: call `canEqual` instead?

File: gobblin-runtime/src/main/java/gobblin/runtime/Task.java [hunk 2]
```diff
@@ -141,6 +143,29 @@
   private final AtomicBoolean shutdownRequested;
   private final CountDownLatch shutdownLatch;
 
+  public Task() {
```
Comment (ydai1124) [inline gobblin-runtime/src/main/java/gobblin/runtime/Task.java]: You can use `@NoArgsConstructor(force = true)`

File: gobblin-runtime/src/main/java/gobblin/runtime/task/TaskIFaceWrapper.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gobblin.runtime.task;
+
+import com.google.common.base.Optional;
```
Comment (ydai1124) [inline gobblin-runtime/src/main/java/gobblin/runtime/task/TaskIFaceWrapper.java]: import order.

File: gobblin-runtime/src/main/java/gobblin/runtime/SafeDatasetCommit.java [hunk 4]
```diff
@@ -129,27 +150,62 @@ public Void call()
...
 
   /**
-   * Synchronized version of {@link #commitDataset(JobState.DatasetState, DataPublisher)} used when publisher is not
+   * Synchronized version of {@link #commitDataset(Collection, DataPublisher)} used when publisher is not
    * thread safe.
    */
-  private synchronized void safeCommitDataset(JobState.DatasetState datasetState, DataPublisher publisher) {
-    commitDataset(datasetState, publisher);
+  private void safeCommitDataset(Collection<TaskState> taskStates, DataPublisher publisher) {
+    synchronized (GLOBAL_LOCK) {
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/SafeDatasetCommit.java]: Publishers are not necessarily thread safe. In line 108 we check whether the publisher is thread-safe, if it is we don't use syncronization, but if it isn't we do.

Previously the synchronization was mistakenly done only within a single dataset, which is a mistake. This just ensures global synchronization when the publisher is not thread safe.

**Comment by ibuenros** [comment]
@ydai1124 thanks for the comments. Addressed them.

**Pull request merged**